Dynamic execution fails non-deterministically with `InterruptedException` on local work branch

### Description of the bug:


When using `--spawn_strategy=dynamic` Bazel occasionally fails with an internal unhandled exception (exit code `37`) in a non-deterministic way. The failure rate is around 2-3% of invocations and retrying immediately never reproduces the same failure.

### The internal exception appears to only come from the `DynamicSpawnStrategy` class
After some initial investigation I temporarily added the `--debug_spawn_scheduler` flag to get more details. With that flag enabled I was able to consistently get details like this on these failures:
```
INFO: Caught InterruptedException from ExecutionException for local branch of {{Action}}, which may cause a crash.
INFO: CancellationException of remote branch of {{Action}}, returning null
...
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.AssertionError: Neither branch of {{Action}} completed. Local was not cancelled and done and remote was cancelled and done.
```

### The issue does not appear related to the `--dynamic_local_strategy`
The `{{Action}}` referenced in the error here is always different. Sometimes it comes from a `KotlinCompile` action which uses `--dynamic_local_strategy=KotlinCompile=worker`, and other times it comes from `JavaIJar` actions which use `--dynamic_local_strategy=JavaIjar=local`. So I don't think the issue is specific to the local strategy used. Although we have not observed any actions using the `sandboxed` strategy failing in this way. That may just be due to how few actions run with the `sandboxed` strategy in our builds though. We have also seen rare cases of this failure happening on 'external' rules like `external/io_bazel_rules_scala_scala_library_2_13_12/io_bazel_rules_scala_scala_library_2_13_12.stamp/scala-library-2.13.12-stamped.jar`.

### The issue does not appear to be related to machine resource utilization
Before the investigation began, my initial assumption was that this issue was caused by the OOM killer or some other resource limitation on the current machine. However after digging through our analytics during these builds it seems extremely unlikely that this is the case. We've seen such issues in the past on smaller machines, but aside from failing in a very different way, the actual utilization on these machines is relatively low. On the couple dozen cases that I've spot checked CPU, RAM, Disk & Network utilization were all well below 50%. Even zooming in the time window to the minute or two before failure doesn't show any spike in utilization or any hint of throttling.

### The issue DOES appear related to the remote cache
One commonality I saw when going through the detailed profiles for these failures was that the action that failed shouldn't be affected by the changes present on that commit. That is to say that we would have expected that action to be cached remotely already. At this point it's worth mentioning that these builds are being run as a part of our CI pipeline, and that machines are assigned arbitrarily. This means that it is very normal for a single machine to run builds on different commits, jumping back and forth in time as old and new commits are tested. So part of me suspects that this issue only really reproduces when the output base is mostly invalidated due to jumping between very different commits. To potentially mitigate this I tried enabling disk cache. At first I thought this mitigation helped, but over time we still saw this failure scenario. Although due to how these CI machines are scaled based off of demand, it's possible we just aren't able to keep the disk cache warm enough to effectively mitigate.

### Which category does this issue belong to?

Core, Local Execution, Remote Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


I have not been able to reproduce this issue in any artificial scenario:
- I have tried some suggestions like `--experimental_local_execution_delay=0` & `--jobs=32` to try to force more actions to actively race on the dynamic strategy.
- I have tried entirely deleting the output base in-between runs.
- I have tried cherry-picking specific invocations we observed this failure on and running them verbatim
- I have tried with and without disk cache
- I have tried all combinations and permutations of the above

### If I cannot reproduce the issue, how confident am I about the details here?
I'm very confident that this issue is real. Despite not being able to artificially reproduce this behavior, and claiming a relatively low failure rate, this issue is very detectable in my systems. As mentioned above this issue is observed in the CI pipeline at my place of work, so even with the low failure rate I still have significant data to back up this issue. To provide some very rough details:
- I had dynamic execution enabled on 'small builds' in our pipeline over a 3 week period
  - The threshold for a small build is something I configure.
  - Changing this threshold does not substantially impact the failure rate of this issue.
  - We have observed this issue on builds ranging from 10K to 50 targets.
  - I'm not sure if it's statistically significant, but if anything the failure rate was slightly higher on very small builds of <500 targets. (maybe 4% up from 2-3%)
- Over this time period we ran over 20k of these dynamic builds and saw over 500 of these failures
- Due to only being able to reproduce this in our CI pipeline, only small, incremental & safe changes to the build are possible to help diagnose the issue
  - We could only enable `--debug_spawn_scheduler` temporarily due to the volume of log spew it produces
  - We cannot bisect the Bazel version or try potentially unstable patch fixes.

### Which operating system are you running Bazel on?


Linux x86_64

### What is the output of `bazel info release`?


7.4.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


_No response_

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.


Unclear if this is a regression. Cannot bisect due to inability to reproduce the issue consistently. Issue verified on `7.0.2` & `7.4.0`.

### Have you found anything relevant by searching the web?

- https://github.com/bazelbuild/bazel/issues/20888
  - References the same `InterruptedException` we see here.
  - Discussed lots of possible causes, most of which revolve around local persistent workers.
  - We tried using `--experimental_dynamic_ignore_local_signals=8,9,10` as outlined in this thread. It did not appear to have much impact.
- https://github.com/bazelbuild/bazel/issues/22482
  - Possibly related. Not enough context on this issue to be sure though.
- https://github.com/bazelbuild/bazel/issues/21773#issuecomment-2388716250
  - Mentions the `InterruptedException` and implies that an error performing downloads could be the root cause. This aligns with some of the other observations we have made.
  - The specific issue here was fixed in `7.4.0` though, and we have not seen this issue fixed with that upgrade.

### Any other information, logs, or outputs that you want to share?

### An abridged list of flags enabled on these builds
```
--noenable_bzlmod
--nolegacy_important_outputs
--watchfs
--experimental_repository_cache_hardlinks
--incompatible_allow_tags_propagation
--experimental_guard_against_concurrent_changes
--noslim_profile
--noexperimental_merged_skyframe_analysis_execution
--java_language_version=17
--java_runtime_version=remotejdk_17
--tool_java_language_version=17
--tool_java_runtime_version=remotejdk_17
--javacopt="-XepDisableAllChecks --release 8"
--host_javacopt=-XepDisableAllChecks
--experimental_java_classpath=off
--nojava_header_compilation
--noincompatible_java_common_parameters
--sandbox_block_path=/usr/local
--sandbox_block_path=/opt
--verbose_failures
--experimental_profile_include_target_label
--experimental_collect_system_network_usage
--host_cxxopt=-std=c++14
--experimental_repository_cache_urls_as_default_canonical_id
--incompatible_enable_proto_toolchain_resolution
--local_cpu_resources=HOST_CPUS
--local_ram_resources=HOST_RAM
--remote_upload_local_results
--remote_timeout=3600
--jobs=1536
--define=EXECUTOR=remote
--remote_default_exec_properties=OSFamily=Linux
--grpc_keepalive_time=30s
--noexperimental_throttle_action_cache_check
--action_env=BAZEL_DO_NOT_DETECT_CPP_TOOLCHAIN=1
--incompatible_strict_action_env
--incompatible_enable_cc_toolchain_resolution
--extra_execution_platforms=@aspect_gcc_toolchain//platforms:x86_64_linux_remote
--host_platform=@aspect_gcc_toolchain//platforms:x86_64_linux_remote
--symlink_prefix=dist/
--crosstool_top=@gcc_toolchain_x86_64//:_cc_toolchain
--host_cpu=k8
--cpu=k8
--experimental_remote_mark_tool_inputs
--heap_dump_on_oom
--internal_spawn_scheduler
--experimental_worker_cancellation
--noworker_multiplex
--worker_quit_after_build
--dynamic_local_strategy=sandboxed,local,worker
--experimental_dynamic_local_load_factor=2
--local_termination_grace_seconds=60
--remote_download_outputs=all
--remote_local_fallback
--spawn_strategy=dynamic
--dynamic_local_strategy=Javac=worker
--dynamic_local_strategy=JavaDeployJar=local
--dynamic_local_strategy=JavaIjar=local
--dynamic_local_strategy=JavaSourceJar=local
--dynamic_local_strategy=KotlinCompile=worker
--dynamic_local_strategy=Turbine=local
--disk_cache=~/.cache/_bazel_disk_cache
--experimental_disk_cache_gc_max_size=500G
--experimental_disk_cache_gc_max_age=0
--experimental_disk_cache_gc_idle_delay=1m
--nobuild_runfile_links
--strategy=TestRunner=remote
--strategy=Genrule=remote
--experimental_dynamic_ignore_local_signals=8,9,10
--bes_upload_mode=FULLY_ASYNC
```

--- a/src/main/java/com/google/devtools/build/lib/dynamic/Branch.java
+++ b/src/main/java/com/google/devtools/build/lib/dynamic/Branch.java
@@ -15,6 +15,7 @@
 package com.google.devtools.build.lib.dynamic;
 
 import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.flogger.GoogleLogger;
@@ -48,8 +49,10 @@ abstract class Branch implements Callable<ImmutableList<SpawnResult>> {
    * #callImpl(ActionExecutionContext)} yet.
    */
   protected final AtomicBoolean starting = new AtomicBoolean(true);
+
   /** The {@link Spawn} this branch is running. */
   protected final Spawn spawn;
+
   /**
    * The {@link SettableFuture} with the results from running the spawn. Must not be null if
    * execution succeeded.
@@ -61,6 +64,7 @@ abstract class Branch implements Callable<ImmutableList<SpawnResult>> {
    * This object is shared between the local and remote branch of an action.
    */
   protected final AtomicReference<DynamicMode> strategyThatCancelled;
+
   /**
    * Semaphore that indicates whether this branch is done, i.e. either completed or cancelled. This
    * is needed to wait for the branch to finish its own cleanup (e.g. terminating subprocesses) once
@@ -70,6 +74,8 @@ abstract class Branch implements Callable<ImmutableList<SpawnResult>> {
 
   protected final DynamicExecutionOptions options;
   protected final ActionExecutionContext context;
+
+  protected Branch otherBranch;
 
   /**
    * Creates a new branch of dynamic execution.
@@ -128,6 +134,41 @@ abstract class Branch implements Callable<ImmutableList<SpawnResult>> {
   /** Executes this branch using the provided executor. */
   public void execute(ListeningExecutorService executor) {
     future.setFuture(executor.submit(this));
+  }
+
+  /** Sets up the {@link Future} used in the current branch to know what other branch to cancel. */
+  protected void prepareFuture(Branch otherBranch) {
+    this.otherBranch = otherBranch;
+    future.addListener(
+        () -> {
+          if (starting.compareAndSet(true, false)) {
+            // If the current branch got cancelled before even starting, we release its semaphore
+            // for it.
+            done.release();
+          }
+          // If the current branch succeeds, there is no need to keep the other branch running.
+          // If the current branch fails, cancel the other branch as well. However, that one may
+          // in turn cancel us, thus causing an interruption. Don't consider that a failure as
+          // we otherwise risk canceling both branches.
+          var state = future.state();
+          if (state == Future.State.SUCCESS
+              || (state == Future.State.FAILED
+                  && !(future.exceptionNow() instanceof InterruptedException))) {
+            otherBranch.cancel();
+          }
+          if (options.debugSpawnScheduler) {
+            logger.atInfo().log(
+                "In listener callback, the future of the remote branch is %s",
+                future.state().name());
+            try {
+              future.get();
+            } catch (InterruptedException | ExecutionException e) {
+              logger.atInfo().withCause(e).log(
+                  "The future of the remote branch failed with an exception.");
+            }
+          }
+        },
+        directExecutor());
   }
 
   /**

--- a/src/main/java/com/google/devtools/build/lib/dynamic/DynamicSpawnStrategy.java
+++ b/src/main/java/com/google/devtools/build/lib/dynamic/DynamicSpawnStrategy.java
@@ -54,6 +54,7 @@ import java.util.Optional;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future.State;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Function;
@@ -650,7 +651,7 @@ public class DynamicSpawnStrategy implements SpawnStrategy {
             // This can happen if the other branch is local under local_lockfree and has returned
             // its result but not yet cancelled this branch, or if the other branch was already
             // cancelled for other reasons. In the latter case, we are good to continue.
-            if (!otherBranch.isCancelled()) {
+            if (otherBranch.future.state() == State.SUCCESS) {
               throw new DynamicInterruptedException(
                   String.format(
                       "Execution of %s strategy stopped because %s strategy could not be cancelled",

--- a/src/main/java/com/google/devtools/build/lib/dynamic/LocalBranch.java
+++ b/src/main/java/com/google/devtools/build/lib/dynamic/LocalBranch.java
@@ -20,7 +20,6 @@ import static com.google.devtools.build.lib.actions.DynamicStrategyRegistry.Dyna
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 import com.google.common.flogger.GoogleLogger;
-import com.google.common.util.concurrent.MoreExecutors;
 import com.google.devtools.build.lib.actions.ActionExecutionContext;
 import com.google.devtools.build.lib.actions.DynamicStrategyRegistry;
 import com.google.devtools.build.lib.actions.DynamicStrategyRegistry.DynamicMode;
@@ -29,6 +28,7 @@ import com.google.devtools.build.lib.actions.SandboxedSpawnStrategy;
 import com.google.devtools.build.lib.actions.Spawn;
 import com.google.devtools.build.lib.actions.SpawnResult;
 import com.google.devtools.build.lib.actions.SpawnResult.Status;
+import com.google.devtools.build.lib.actions.SpawnStrategy;
 import com.google.devtools.build.lib.dynamic.DynamicExecutionModule.IgnoreFailureCheck;
 import com.google.devtools.build.lib.profiler.Profiler;
 import com.google.devtools.build.lib.profiler.SilentCloseable;
@@ -36,7 +36,6 @@ import com.google.devtools.build.lib.util.io.FileOutErr;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.Optional;
-import java.util.concurrent.ExecutionException;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Function;
@@ -50,7 +49,6 @@ import javax.annotation.Nullable;
 class LocalBranch extends Branch {
   private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
 
-  private RemoteBranch remoteBranch;
   private final IgnoreFailureCheck ignoreFailureCheck;
   private final Function<Spawn, Optional<Spawn>> getExtraSpawnForLocalExecution;
   private final AtomicBoolean delayLocalExecution;
@@ -139,38 +137,12 @@ class LocalBranch extends Branch {
     throw new AssertionError("canExec passed but no usable local strategy for action " + spawn);
   }
 
-  /** Sets up the {@link Future} used in the local branch to know what remote branch to cancel. */
-  protected void prepareFuture(RemoteBranch remoteBranch) {
-    // TODO(b/203094728): Maybe generify this method and move it up.
-    this.remoteBranch = remoteBranch;
-    future.addListener(
-        () -> {
-          if (starting.compareAndSet(true, false)) {
-            // If the local branch got cancelled before even starting, we release its semaphore
-            // for it.
-            done.release();
-          }
-          if (!future.isCancelled()) {
-            remoteBranch.cancel();
-          }
-          if (options.debugSpawnScheduler) {
-            logger.atInfo().log(
-                "In listener callback, the future of the local branch is %s",
-                future.state().name());
-            try {
-              future.get();
-            } catch (InterruptedException | ExecutionException e) {
-              logger.atInfo().withCause(e).log(
-                  "The future of the local branch failed with an exception.");
-            }
-          }
-        },
-        MoreExecutors.directExecutor());
-  }
-
   @Override
   ImmutableList<SpawnResult> callImpl(ActionExecutionContext context)
       throws InterruptedException, ExecException {
+    if (otherBranch == null) {
+      throw new IllegalStateException("prepareFuture not called");
+    }
     try {
       if (!starting.compareAndSet(true, false)) {
         // If we ever get here, it's because we were cancelled early and the listener
@@ -191,7 +163,7 @@ class LocalBranch extends Branch {
               maybeIgnoreFailure(exitCode, errorMessage, outErr);
             }
             DynamicSpawnStrategy.stopBranch(
-                remoteBranch, this, strategyThatCancelled, options, this.context);
+                otherBranch, this, strategyThatCancelled, options, this.context);
           },
           getExtraSpawnForLocalExecution);
     } catch (DynamicInterruptedException e) {

--- a/src/main/java/com/google/devtools/build/lib/dynamic/RemoteBranch.java
+++ b/src/main/java/com/google/devtools/build/lib/dynamic/RemoteBranch.java
@@ -20,7 +20,6 @@ import static com.google.devtools.build.lib.actions.DynamicStrategyRegistry.Dyna
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 import com.google.common.flogger.GoogleLogger;
-import com.google.common.util.concurrent.MoreExecutors;
 import com.google.devtools.build.lib.actions.ActionExecutionContext;
 import com.google.devtools.build.lib.actions.DynamicStrategyRegistry;
 import com.google.devtools.build.lib.actions.DynamicStrategyRegistry.DynamicMode;
@@ -29,10 +28,10 @@ import com.google.devtools.build.lib.actions.SandboxedSpawnStrategy;
 import com.google.devtools.build.lib.actions.SandboxedSpawnStrategy.StopConcurrentSpawns;
 import com.google.devtools.build.lib.actions.Spawn;
 import com.google.devtools.build.lib.actions.SpawnResult;
+import com.google.devtools.build.lib.actions.SpawnStrategy;
 import com.google.devtools.build.lib.dynamic.DynamicExecutionModule.IgnoreFailureCheck;
 import com.google.devtools.build.lib.events.Event;
 import com.google.devtools.build.lib.util.io.FileOutErr;
-import java.util.concurrent.ExecutionException;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
@@ -45,7 +44,6 @@ import javax.annotation.Nullable;
 class RemoteBranch extends Branch {
   private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
 
-  private LocalBranch localBranch;
   private final IgnoreFailureCheck ignoreFailureCheck;
   private final AtomicBoolean delayLocalExecution;
 
@@ -109,39 +107,11 @@ class RemoteBranch extends Branch {
     throw new AssertionError("canExec passed but no usable remote strategy for action " + spawn);
   }
 
-  /** Sets up the future for this branch, once the other branch is available. */
-  public void prepareFuture(LocalBranch localBranch) {
-    this.localBranch = localBranch;
-    future.addListener(
-        () -> {
-          if (starting.compareAndSet(true, false)) {
-            // If the remote branch got cancelled before even starting, we release its semaphore
-            // for it.
-            done.release();
-          }
-          if (!future.isCancelled()) {
-            localBranch.cancel();
-          }
-          if (options.debugSpawnScheduler) {
-            logger.atInfo().log(
-                "In listener callback, the future of the remote branch is %s",
-                future.state().name());
-            try {
-              future.get();
-            } catch (InterruptedException | ExecutionException e) {
-              logger.atInfo().withCause(e).log(
-                  "The future of the remote branch failed with an exception.");
-            }
-          }
-        },
-        MoreExecutors.directExecutor());
-  }
-
   @Override
   public ImmutableList<SpawnResult> callImpl(ActionExecutionContext context)
       throws InterruptedException, ExecException {
-    if (localBranch == null) {
-      throw new IllegalStateException("Initialize not called");
+    if (otherBranch == null) {
+      throw new IllegalStateException("prepareFuture not called");
     }
     try {
       if (!starting.compareAndSet(true, false)) {
@@ -158,7 +128,7 @@ class RemoteBranch extends Branch {
               maybeIgnoreFailure(exitCode, errorMessage, outErr);
             }
             DynamicSpawnStrategy.stopBranch(
-                localBranch, this, strategyThatCancelled, options, this.context);
+                otherBranch, this, strategyThatCancelled, options, this.context);
           },
           delayLocalExecution);
     } catch (DynamicInterruptedException e) {

--- a/src/test/java/com/google/devtools/build/lib/dynamic/DynamicSpawnStrategyTest.java
+++ b/src/test/java/com/google/devtools/build/lib/dynamic/DynamicSpawnStrategyTest.java
@@ -1091,6 +1091,177 @@ public class DynamicSpawnStrategyTest {
     assertThatStrategyPropagatesException(localExec, remoteExec, e);
   }
 
+  @Test
+  public void remoteBranchThrowsInterruptedExceptionAfterCancellation() throws Exception {
+    CountDownLatch localCanFinish = new CountDownLatch(1);
+    CountDownLatch remoteStarted = new CountDownLatch(1);
+
+    MockSpawnStrategy localStrategy =
+        new MockSpawnStrategy(
+            "MockLocalSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              remoteStarted.await();
+              localCanFinish.await();
+            },
+            DoExec.NOTHING);
+
+    MockSpawnStrategy remoteStrategy =
+        new MockSpawnStrategy(
+            "MockRemoteSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              remoteStarted.countDown();
+              Thread.sleep(TestUtils.WAIT_TIMEOUT_MILLISECONDS);
+            },
+            DoExec.NOTHING);
+
+    StrategyAndContext strategyAndContext = createSpawnStrategy(localStrategy, remoteStrategy);
+
+    TestThread testThread =
+        new TestThread(
+            () -> {
+              try {
+                Spawn spawn = newDynamicSpawn();
+                strategyAndContext.exec(spawn);
+              } catch (InterruptedException | ExecException e) {
+                throw new RuntimeException(e);
+              }
+            });
+    testThread.start();
+
+    remoteStarted.await();
+    localCanFinish.countDown();
+    testThread.joinAndAssertState(TestUtils.WAIT_TIMEOUT_MILLISECONDS);
+
+    assertThat(localStrategy.succeeded()).isTrue();
+    assertThat(remoteStrategy.succeeded()).isFalse();
+  }
+
+  @Test
+  public void localBranchCancelsRemoteWhichThrowsInterruptedException() throws Exception {
+    CountDownLatch bothStarted = new CountDownLatch(2);
+
+    MockSpawnStrategy localStrategy =
+        new MockSpawnStrategy(
+            "MockLocalSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              countDownAndWait(bothStarted);
+            },
+            DoExec.NOTHING);
+
+    MockSpawnStrategy remoteStrategy =
+        new MockSpawnStrategy(
+            "MockRemoteSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              countDownAndWait(bothStarted);
+              Thread.sleep(TestUtils.WAIT_TIMEOUT_MILLISECONDS);
+            },
+            DoExec.NOTHING);
+
+    StrategyAndContext strategyAndContext = createSpawnStrategy(localStrategy, remoteStrategy);
+
+    Spawn spawn = newDynamicSpawn();
+    strategyAndContext.exec(spawn);
+
+    assertThat(localStrategy.succeeded()).isTrue();
+    assertThat(remoteStrategy.succeeded()).isFalse();
+    assertThat(outErr.outAsLatin1()).contains("output files written with MockLocalSpawnStrategy");
+  }
+
+  @Test
+  public void remoteFinishesFirstAndThrowsInterruptedExceptionDuringLocalCancellation()
+      throws Exception {
+    CountDownLatch bothStarted = new CountDownLatch(2);
+
+    MockSpawnStrategy localStrategy =
+        new MockSpawnStrategy(
+            "MockLocalSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              countDownAndWait(bothStarted);
+              Thread.sleep(TestUtils.WAIT_TIMEOUT_MILLISECONDS);
+            },
+            DoExec.NOTHING);
+
+    MockSpawnStrategy remoteStrategy =
+        new MockSpawnStrategy(
+            "MockRemoteSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              countDownAndWait(bothStarted);
+            },
+            DoExec.NOTHING);
+
+    StrategyAndContext strategyAndContext = createSpawnStrategy(localStrategy, remoteStrategy);
+
+    Spawn spawn = newDynamicSpawn();
+    // This should complete successfully with remote results, not throw AssertionError
+    strategyAndContext.exec(spawn);
+
+    assertThat(localStrategy.succeeded()).isFalse();
+    assertThat(remoteStrategy.succeeded()).isTrue();
+    assertThat(outErr.outAsLatin1()).contains("output files written with MockRemoteSpawnStrategy");
+  }
+
+  @Test
+  public void remoteBranchThrowsInterruptedExceptionBeforeStopConcurrentSpawns() throws Exception {
+    CountDownLatch remoteStarted = new CountDownLatch(1);
+
+    MockSpawnStrategy localStrategy =
+        new MockSpawnStrategy(
+            "MockLocalSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              remoteStarted.await();
+              Thread.sleep(100); // Give remote time to fail
+            },
+            DoExec.NOTHING);
+
+    MockSpawnStrategy remoteStrategy =
+        new MockSpawnStrategy(
+            "MockRemoteSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              remoteStarted.countDown();
+              throw new InterruptedException("Remote operation interrupted");
+            },
+            DoExec.NOTHING);
+
+    StrategyAndContext strategyAndContext = createSpawnStrategy(localStrategy, remoteStrategy);
+
+    Spawn spawn = newDynamicSpawn();
+    strategyAndContext.exec(spawn);
+
+    assertThat(localStrategy.succeeded()).isTrue();
+    assertThat(remoteStrategy.succeeded()).isFalse();
+    assertThat(outErr.outAsLatin1()).contains("output files written with MockLocalSpawnStrategy");
+  }
+
+  @Test
+  public void remoteBranchInterruptedAfterLocalStartsStopConcurrentSpawns() throws Exception {
+    CountDownLatch bothStarted = new CountDownLatch(2);
+    CountDownLatch localReadyToStop = new CountDownLatch(1);
+
+    MockSpawnStrategy localStrategy =
+        new MockSpawnStrategy(
+            "MockLocalSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> countDownAndWait(bothStarted),
+            (self, spawn, actionExecutionContext) -> localReadyToStop.countDown());
+
+    MockSpawnStrategy remoteStrategy =
+        new MockSpawnStrategy(
+            "MockRemoteSpawnStrategy",
+            (self, spawn, actionExecutionContext) -> {
+              countDownAndWait(bothStarted);
+              Thread.sleep(TestUtils.WAIT_TIMEOUT_MILLISECONDS);
+            },
+            DoExec.NOTHING);
+
+    StrategyAndContext strategyAndContext = createSpawnStrategy(localStrategy, remoteStrategy);
+
+    Spawn spawn = newDynamicSpawn();
+    strategyAndContext.exec(spawn);
+
+    assertThat(localStrategy.succeeded()).isTrue();
+    assertThat(remoteStrategy.succeeded()).isFalse();
+    assertThat(outErr.outAsLatin1()).contains("output files written with MockLocalSpawnStrategy");
+  }
+
   private static FailureDetail createFailureDetail(String message) {
     return FailureDetail.newBuilder()
         .setMessage(message)

